python/ray/experimental/tf_utils.py

```diff
@@ -196,3 +199,421 @@ def set_weights(self, new_weights):
+                    tf_constants.append(processed_args[-1])
+                    tensors.append(processed_args[-1])
+                elif arg_type == "native":
+                    processed_args.append(arg)
+                else:
+                    raise ValueError("Invalid argument type: {}".format(arg_type))
+
+            # process the kwargs
+            processed_kwargs = {}
+            for arg_type, arg_dtype, arg, kw in zip(kwarg_types, kwarg_dtypes, kwargs, kws):
+                if arg_type == "tf_var":
+                    processed_kwargs[kw] = tf.Variable(arg, dtype=arg_dtype)
+                    tensors.append(processed_kwargs[kw])
+                elif arg_type == "tf_const":
+                    processed_kwargs[kw] = tf.constant(arg, dtype=arg_dtype)
+                    tf_constants.append(processed_kwargs[kw])
+                    tensors.append(processed_kwargs[kw])
+                elif arg_type == "native":
+                    processed_kwargs[kw] = arg
+                else:
+                    raise ValueError("Invalid keyword argument type: {}".format(arg_type))
+
+            # execute internal method
+            with tf.GradientTape(persistent=persistent_tape) as tape:
+                # watch all of the constants 
+                for constant in tf_constants:
+                    tape.watch(constant)
+
+                results = method(self, *processed_args, **processed_kwargs)
+
+            # cache the TF state for backward pass
+            if not hasattr(self, "__ray_tf_info__"):
+                self.__ray_tf_info__ = {}
+            self.__ray_tf_info__[identifier] = (tape, tensors, results)
+
+            # process the results
+            if isinstance(results, (list, tuple)):
+                processed_results = [result.numpy() if isinstance(result, (tf.Tensor, tf.Variable)) 
+                                     else result for result in results]
+            else:
+                processed_results = results.numpy() if isinstance(results, (tf.Tensor, tf.Variable)) else results
+
+            return processed_results
+        else:
+            assert kwarg_types is None and kwarg_dtypes is None and kws is None, ("During backwards pass, kwarg_types, kwarg_dtypes, "
+                                                                                  "and kws must be None but are: "
+                                                                                  "{}, {}, and {}, respectively.".format(kwarg_types, kwarg_dtypes, kws)) 
+
+            # fetch the cached TF state
+            tape, tensors, results = self.__ray_tf_info__[identifier]
+
+            # process the partial downstream gradients
+            if not isinstance(results, (list, tuple)):
+                results = [results]
+            dys_processed = [] 
+            for dy, result in zip(dys, results):
+                if isinstance(dy, np.floating) and dy == 0.0:
```

Comment:
It can happen in the case of a literal `0.0` gradient, right? If you mean detecting whether or not a `0.0` appeared between the downstream op's `grad` and the current `grad` that's a trickier problem.

Things do break without the `if` statement, but that's because shapes don't match up and a `reshape` op fails in TF.